### DISCUSSION TITLE
PSA: Emacs crash with Enchant-2 version 2.7.1 on MacOS

### DISCUSSION THREAD

**Discussion by jeff-phil** [minad/jinx#167] [General]
Just a head's up, in case anyone else has issue.

On my mac, I upgraded to Enchant-2 version 2.7.1 and now Emacs crashes when jinx-mode is enabled.

I opened a new issue in their repo: https://github.com/AbiWord/enchant/issues/370 with details.

**Comment by minad** [comment]
Thanks! It's already fixed 🎉

**Comment by jeff-phil** [comment]
Yes, that is fabulous.

Not sure if noticed mentioned that was because passing -1 for the `len` parameter to `enchant_dict_check` for the word to be checked. 

I see their [doc comments](https://github.com/AbiWord/enchant/blob/3076ee45ebd12bc9246c11bbb45772c032f72fa4/lib/enchant.h#L169) says to pass -1 to have enchant calculate strlen (and for a few others like suggest and add). But wonder if better for long-term compatibility to just go ahead and calculate the size in jinx-mod.c and send, regardless of comments.  Thoughts?

**Reply by minad** [reply]
It doesn't seem to be a common issue, since I've not seen any crashes so far with -1 and it is documented like this. Is the crash due to a new backend?

**Reply by jeff-phil** [reply]
> Is the crash due to a new backend?

Yes, from this being rewritten in Vala: https://github.com/AbiWord/enchant/issues/366, and related to a newly converted code.

> It doesn't seem to be a common issue, since I've not seen any crashes so far with -1 and it is documented like this.

Yep, agree.  And the maintainer there was surprised being called with -1.   🤷‍♂️ 

I was more thinking a simple` strlen` fix in 3 spots of jinx-mod.c to send the actual size would be more idiomatic when passing strings. I'm a bit surprised enchant ever allowed to pass -1, since seems like could be a potential issue security or otherwise.

But fix was fast, so I'll close this discussion.

**Reply by minad** [reply]
> I was more thinking a simple strlen fix in 3 spots of jinx-mod.c to send the actual size would be more idiomatic when passing strings. I'm a bit surprised enchant ever allowed to pass -1, since seems like could be a potential issue security or otherwise.

Passing a length is mainly useful if you want to limit the size of the checked string (if the string is not \\0 terminated). Supporting -1 (or even other negative values) as special value seems like perfectly good C API to me. The Emacs string conversion always nul terminates the string. However there may be cases with inner nul characters. Technically these can probably exist, but we won't have problems here since words in natural languages usually don't include nul.

**Comment by jeff-phil** [comment]
Fix completed with v2.7.2 of enchant.
